src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -3990,16 +3938,17 @@ private enum PostSkipAction
 
         private PostSkipAction SkipBadTokensWithExpectedKind(
             Func<LanguageParser, bool> isNotExpectedFunction,
-            Func<LanguageParser, bool> abortFunction,
+            Func<LanguageParser, SyntaxKind, bool> abortFunction,
             SyntaxKind expected,
+            SyntaxKind closeKind,
             out GreenNode trailingTrivia)
         {
             var nodes = _pool.Allocate();
             bool first = true;
             var action = PostSkipAction.Continue;
             while (isNotExpectedFunction(this))
             {
-                if (abortFunction(this))
+                if (abortFunction(this, closeKind) || this.IsTerminator())
```

Comment:
this.IsTerminator() is common in all abort functions, so it got extracted from there and moved here.